Add flag to administratively enable APIs on bootstrap

joestringer opened the pull request:
Depends on ~https://github.com/cilium/cilium/pull/24967~ .

Add a new (hidden) flag, `enable-cilium-api-server-access`, which accepts a
list of APIs that the daemon should allow REST API clients to mutate.

This flag can be useful to restrict the sorts of changes that can be
made locally directly from the node or from users with access to the
Cilium container. Many of these have been added over time to assist
development and debugging purposes, and may not be necessary in
production environments.

Set the default to allow "all" API modifications (`*`) so that there is no
change in behaviour for existing deployments.

In order to successfully use this flag, it is likely that at least the
following options must be configured in order to allow cilium-cni to provision
pods, and to ensure that the health of the agent can be assessed by Kubernetes:

- `GetConfig`
- `GetHealthz`
- `PutEndpointID`
- `DeleteEndpointID`
- `PostIPAM`
- `DeleteIPAMIP`

Users may additionally see benefit in providing `Get*` as a wildcard in order
to allow any read-only APIs to be accessed, which can be very useful for
debugging.

Occasionally, additional "flush" (Delete) APIs can also be very useful in order
to mitigate issues at runtime, so this option should not be used unless you
have a pretty strong understanding of the APIs you are likely to rely upon.

For completeness, similar flags are also added for the cilium-health and
operator APIs, though the usefulness there is more limited.

Related docs: https://github.com/cilium/cilium/pull/24968

joestringer committed:
api: Add helper to administratively disable APIs

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
daemon: Add warning for disabled APIs

The newly added API access flags are very flexible in the APIs that they
allow to be administratively disabled, but there are some options that
are really required for any standard Cilium operations. Add some basic
sanity checking before applying the user configuration in these cases,
as it could lead to unexpected results such as inability to keep the
Cilium agent running or inability to deploy new endpoints.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
daemon: Add Cells for Cilium API specifications

This preparatory commit introduces a new Cell for each Swagger API
Specification in order to allow reuse earlier on during the agent
initialization / lifecycle. No functional changes intended.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
daemon: Add flags for API access allowlist

Add new flag(s) to the daemon which restrict the ability for API clients
to call certain API endpoints. The new option is an allowlist of
Pascalized API endpoints that may be allowed, or optionally a Prefix
followed by the '*' character in order to allow a wildcard of API
endpoints, for example "*" for all API endpoints, or "Get*" for all GET
endpoints.

Set the default to allow all ("*") API modifications so that there is no
change in behaviour for existing deployments.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
health: Rename variable to avoid pkg conflict

pkg/api will be used in an upcoming commit, so rename the variable in
this function to avoid that conflict.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
daemon: Add cilium-health API restrictions flag

Similar to the recent commit to administratively enable/disable support for
agent API endpoints, extend this support to the cilium-health API
exposed by nodes as well.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
operator: Rename variable to avoid pkg conflict

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
operator: Add operator API restrictions flag

For now this is not particularly important as the operator only has a
/healthz API to get the health of the operator. However, if we ever
extend this API in future then it could become useful to consistently
apply API restrictions via the new flag.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
api: Add flags to configure API access from spec

Use the autogenerated API specifications in order to populate "allowed
APIs" flags in the api specification objects for each API. This code
will be hooked in for usage in subsequent commits.

Signed-off-by: Joe Stringer <joe@cilium.io>

joamaki commented on pkg/api/apidisable.go:
```diff
@@ -0,0 +1,42 @@
+	wr.WriteHeader(http.StatusForbidden)
+	log.WithFields(logrus.Fields{
+		logfields.Endpoint: a.name,
+	}).Info("Denied API request on admininstratively disabled API endpoint")
```
s/admininstratively/administratively/

joamaki commented on api/v1/server.gotmpl:
```diff
@@ -122,6 +123,28 @@ var (
+	*loads.Document
+}
+
+func newSpec() (*Spec, error) {
```
Please rebase and change the existing `Cell` to also provide the spec, e.g. `newForCell` could return `(*Spec, *Server, error)` and remove `SpecCell`.

joamaki replied:
Ah, after reading rest of the commits I realized we'll be using this in situations where we're not yet using the `server.Cell`. Perhaps then we'll want to change `newForCell` to depend on `*Spec` instead. At least for the time being.

joestringer replied:
Ah yes I missed that during the rebase, good point. I'll fix that up.

joestringer replied:
Diff:
```
@@ -57,11 +58,7 @@ type serverParams struct {                                                                                                                                                                                                                                              
 }                                                                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                                           
 func newForCell(p serverParams) (*Server, error) {                                                                                                                                                                                                                                        
-       swaggerSpec, err := loads.Analyzed(SwaggerJSON, "")                                                                                                                                                                                                                                
-       if err != nil {                                                                                                                                                                                                                                                                    
-               return nil, fmt.Errorf("failed to swagger spec: %w", err)                                                                                                                                                                                                                  
-       }                                                                                                                                     
-       api := {{ .Package }}.New{{pascalize .Name}}API(swaggerSpec)                                                                          
+       api := {{ .Package }}.New{{pascalize .Name}}API(p.Spec)                                                                               
                                                                                                                                              
        // Construct the API from the provided handlers                                                                                       
        {{- $package := .Package }} 
```

(+ autogenerated code, I'll fold it into a new patch)

joamaki commented on api/v1/server.gotmpl:
```diff
@@ -123,25 +123,52 @@ var (
   tlsCACertificate  string
 )
 
+type SpecConfig struct {
```
Could make this `ServerConfig` instead and move server-related options into this at some point?

nathanjsweet reviewed (approved):
LGTM for API

joestringer committed:
api: Reuse newly introduced API spec in API cells

Now that there's dedicated cells to provide the API specs for each API,
each server can now directly depend on those rather than doing their own
swagger spec instantiation.

Signed-off-by: Joe Stringer <joe@cilium.io>

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1988,6 +1988,23 @@ func (d *Daemon) instantiateAPI(swaggerSpec *server.Spec) *restapi.CiliumAPIAPI
+	msg := "Required API option %s is disabled. This may prevent Cilium from operating correctly"
+	hint := "Consider enabling this API in " + server.AdminEnableFlag
+	for _, requiredAPI := range []string{
+		"GetConfig",
+		"GetHealthz",
+		"PutEndpointID",
+		"DeleteEndpointID",
+		"PostIPAM",
+		"DeleteIPAMIP",
```
I would suggest to either leave a comment in the code explaining why each one of these APIs are fundamental and / or leave the reasoning for it in the Warning log message so that users / admins know why it is important.

Also, let's say that the admin removed the liveness probe from the daemonset and therefore doesn't need the `GetHealthz` and so it gets administratively disabled. Should we still print a Warning every time Cilium starts?

joestringer replied:
:+1: comments make sense, I can send a follow-up.

Warnings are not perfect, but this is exactly the sort of reason it's a warning and not an error or fatal message. This is a way to let the administrator know that they're running an unconventional setup, and if they know what they're doing and are happy with the consequences then they can ignore it. If you're that far off the beaten path, then I'm OK with having a warning until someone complains that it's a real issue.

joestringer replied:
https://github.com/cilium/cilium/pull/25170

aanm reviewed (approved):
One small non-blocking feedback.
